hedera-node/src/main/java/com/hedera/services/stream/RecordStreamFileWriter.java

```diff
@@ -0,0 +1,490 @@
+			// this is the first object, we should wait for the first complete window
+			result = false;
+		} else {
+			// if lastConsensusTimestamp and currentConsensusTimestamp are in different periods,
+			// we should start a new file
+			result =
+					getPeriod(lastConsensusTimestamp, logPeriodMs) != getPeriod(currentConsensusTimestamp, logPeriodMs);
+		}
+		// update lastConsensusTimestamp
+		lastConsensusTimestamp = currentConsensusTimestamp;
+		return result;
+	}
+
+	/**
+	 * if recordStreamFile is not null:
+	 * write last runningHash to current file;
+	 * close current file;
+	 * and generate a corresponding signature file
+	 */
+	public void closeCurrentAndSign() {
+		if (recordStreamFileBuilder != null) {
+			// generate file name
+			final var firstTxnTimestamp =
+					recordStreamFileBuilder.getRecordStreamItems(0).getRecord().getConsensusTimestamp();
+			this.file = new File(
+					generateStreamFilePath(
+							Instant.ofEpochSecond(firstTxnTimestamp.getSeconds(), firstTxnTimestamp.getNanos())));
+			this.fileNameShort = file.getName();
+			if (file.exists() && !file.isDirectory()) {
+				LOG.debug(OBJECT_STREAM.getMarker(), "Stream file already exists {}", fileNameShort);
+			} else {
+				try {
+					// write endRunningHash
+					final var endRunningHash = runningHash.getFutureHash().get();
+					recordStreamFileBuilder.setEndObjectRunningHash(toProto(endRunningHash));
+					dosMeta.write(endRunningHash.getValue());
+					LOG.debug(OBJECT_STREAM_FILE.getMarker(), "closeCurrentAndSign :: write endRunningHash {}",
+							endRunningHash);
+
+					// write block number to metadata
+					dosMeta.writeLong(recordStreamFileBuilder.getBlockNumber());
+					LOG.debug(OBJECT_STREAM_FILE.getMarker(), "closeCurrentAndSign :: write block number {}",
+							recordStreamFileBuilder.getBlockNumber());
+				} catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+					LOG.error(EXCEPTION.getMarker(),
+							"closeCurrentAndSign :: Got interrupted when getting endRunningHash for writing {}",
+							fileNameShort, e);
+					return;
+				} catch (IOException e) {
+					Thread.currentThread().interrupt();
+					LOG.warn(EXCEPTION.getMarker(),
+							"closeCurrentAndSign :: IOException when serializing endRunningHash and block number into "
+									+ "metadata", e);
+					return;
+				}
+
+				/* When sidecars are enabled we will need to add more logic at this point:
+					1. We should have a map (?) of the TransactionSidecarRecords with key - the type of sidecar, updated
+					everytime we consume a RSO.
+					2. At this point, we will create a SidecarFile for each present type of sidecar record in the current
+					 block
+					3. We populate the SidecarMetadata of the RecordStreamFile message
+				*/
+
+				try {
+					// create record file
+					stream = new FileOutputStream(file, false);
+					dos = new SerializableDataOutputStream(
+							new BufferedOutputStream(new HashingOutputStream(streamDigest, stream)));
```

Comment:
This is a resource leak as we don't close these when there's an error. `dos` and `stream` don't need to be class variables. Move to method variables in a try/with-resources.

Reply:
[Fixed.](https://github.com/hashgraph/hedera-services/pull/3461/commits/5d8e831b26143ae625c5109de968aafa6cf5f022)